src/Node/ChildProcess/StdIO.purs

```diff
@@ -0,0 +1,206 @@
+  }
+
+-- | Ignore this type and just look at `toStdIoOption`
+type ConfigureStdIoOptionsFn stdInUsed stdOutUsed stdErrUsed ipcUsed =
+  { stdin :: Maybe (StdInChoices -> StdIO stdInUsed)
+  , stdout :: Maybe (StdOutErrChoices -> StdIO stdOutUsed)
+  , stderr :: Maybe (StdOutErrChoices -> StdIO stdErrUsed)
+  , ipc :: Maybe (IpcOption False -> IpcOption ipcUsed)
+  , extra :: Array Foreign
+  }
+  -> { stdin :: Maybe (StdInChoices -> StdIO stdInUsed)
+     , stdout :: Maybe (StdOutErrChoices -> StdIO stdOutUsed)
+     , stderr :: Maybe (StdOutErrChoices -> StdIO stdErrUsed)
+     , ipc :: Maybe (IpcOption False -> IpcOption ipcUsed)
+     , extra :: Array Foreign
+     }
+
+-- Using `pipe` or `overlapped` means we can write to the child process' input, so we use `write` here.
+-- But using `inherit` means we're reusing the process' `stdin`, which is `Readable ()`, so we us `read` there.
+choicesStdin :: StdInChoices
+choicesStdin =
+  { usePipe: pipeWrite
+  , useOverlapped: overlappedWrite
+  , useInherit: inheritRead
+  }
+
+-- Using `pipe` or `overlapped` means we can read the child process' output, so we use `read` here.
+-- But using `inherit` means we're reusing the process' `stdout`/`stderr`, which are `Writeable ()`, so we us `write` there.
+choicesStdOutErr :: StdOutErrChoices
+choicesStdOutErr =
+  { usePipe: pipeRead
+  , useOverlapped: overlappedRead
+  , useInherit: inheritWrite
+  }
+
+-- | Intended usage is below. 
+-- | ```
+-- | `spawn' "cmd" [ "args" ] (_ 
+-- |    { stdio = Just $ 
+-- |        toStdIoOption (_ 
+-- |            { stdin = Just inherit
+-- |            , stdout = Just $ shareStream someWriteStream
+-- |            , stderr = Just ignore
+-- |            }
+-- |        )
+-- |    })
+-- | ```
+-- |
+-- | The verbosity of the `ConfigureStdIoOptionsFn` type is needed
+-- | to make `stdin`/`stdout`/`stderr` type safe.
+toStdIoOption
+  :: forall stdInUsed stdOutUsed stdErrUsed ipcUsed
+   . ConfigureStdIoOptionsFn stdInUsed stdOutUsed stdErrUsed ipcUsed
+  -> StdIoOption stdInUsed stdOutUsed stdErrUsed ipcUsed
+toStdIoOption buildOptions = do
+  let
+    stdinOpt = maybe undefined stdIoToForeign $ map (\fn -> fn choicesStdin) options.stdin
+    stdoutOpt = maybe undefined stdIoToForeign $ map (\fn -> fn choicesStdOutErr) options.stdout
+    stderrOpt = maybe undefined stdIoToForeign $ map (\fn -> fn choicesStdOutErr) options.stderr
+    ipcOpt
+      | Just fn <- options.ipc
+      , IpcOption true <- fn noIpc = unsafeToForeign "ipc"
+      | otherwise = undefined
+
+    toStdIO :: Array Foreign -> StdIoOption stdInUsed stdOutUsed stdErrUsed ipcUsed
+    toStdIO = unsafeCoerce
+  toStdIO $ [ stdinOpt, stdoutOpt, stderrOpt, ipcOpt ] <> options.extra
+  where
+  options = buildOptions defaults
+
+  defaults =
+    { stdin: Nothing
+    , stdout: Nothing
+    , stderr: Nothing
+    , ipc: Nothing
+    , extra: []
+    }
+
+  stdIoToForeign :: forall r. StdIO r -> Foreign
+  stdIoToForeign = unsafeToForeign
```

Comment:
A LOT is going on in this function, so let me clarify things for reviewers.

Goals:
1. We want to track at the type level whether a `stdin`/`stdout`/`stderr` stream (1) exists and (2) what can be done with it. We further want to track whether the `"ipc"` option was enabled, which enables the `send` function among other things.
2. We also want these options to be easily overridden by the end-user. In other words, I must enjoy writing the options as an end-user.
3. We want to prevent hard-to-understand compiler errors or at least make it obvious what went wrong.
4. We also want to track whether the `ipc` stdio option was included in the array. This option can only appear at most once in the array. That's why it's handled separately.
5. In the event that the user provides their own streams in `stdio` (hence the `extra` field), we want their access to such things to be consistent.

In pursuit of goal 2, I followed the "defaults" syntax. (e.g. `someFunction $ defaults { label = override }`. However, the `defaults` value is often already known and pointless. So, rather than requiring a value, we require a function that passes that default value into the function. This saves us a few characters and removes the need to import something:
```diff
-someFunction $ defaults { label = value }
+someFunction $ _ { label = value }
```

In pursuit of goal 1... The `stdin`/`stdout`/`stderr` are consumers. We want to get a `Stream x` from the child process (e.g. `Readable ()` for `stdin`). Thus, we need to track on the `ChildProcess` type what the stream type will be for `stdin`/`stdout`/`stderr`. If that value ends up being `Stream ()`, there's nothing anyone can do with that. So, it acts like a type-save `null`. Similarly, if it's `Stream (read :: Proxy "some message")`, one can't do anything with that either (more on why there's a `Proxy` there later). Thus, the value returned by this function stores what the `x` in `Stream x` must be for `stdin`/`stdout`/`stderr`.

If we use `inherit` for `stdio[0]`/`stdin`, the stream we get back is the process' `stdin`, which is a `Readable ()`. If we use `inherit` for `stdio[1]`/`stdout`, we get back the Process' `stdout`, which is a `Writeable ()`. To overload both possibilities and only need to use one value `inherit`, so that it matches Node's docs but is still type safe, we use a callback function.
```
-- type-annotated value for 'inherit'
inheritRead :: StdIO (read :: Read)
inheritRead = unsafeCoerce "inherit"

inheritWrite :: StdIO (write :: Write)
inheritWrite = unsafeCoerce "inherit"

-- select the choice from a record of options
inherit :: forall r. { useInherit :: StdIO x | r } -> StdIO x
inherit { useInherit } = useInherit

type StdInChoices =
  { ... other choices
  , useInherit :: StdIO (read :: Read)
  }
  
type StdOutErrChoices =
  { ... other choices
  , useInherit :: StdIO (write :: Write)
  }
```
For `stdin`, we pass in a record of `StdInChoices`, which ensures that `inherit` will select the `(read :: Read)` row. For `stdout`, we pass in the `StdOutErrChoices`, where `inherit` will select the `(write :: Write)` row. The reason why the row polymorphism is needed here is because some streams that work in `stdin` context (e.g. read-only or read-write streams) don't work in `stdout` context (e.g write-only or read-write streams).

For goal 3, we use monomorphic records for `StdInChoices`/`StdOutErrChoices` to prevent the need to use `RowList` (which has difficulties with polymorphic record inference). This prevents some possible errors.

Lastly, for goal 3, `ignore` will provide a row of `(read :: Proxy "some msg", write :: Proxy "some msg")`. In the event that someone still tries to use `write (stdin childProcess) ...` when they set `stdin` to `ignore`, the message they will get will be a type unification between
```
read :: Read
vs
read :: Proxy "'ignore' was used for this IO slot, so this stream on the child process doesn't exist."
```
which should make it obvious what went wrong. Feedback on what would be a good error message would be helpful as well as where else I should replace a row of `()` with `(read :: Proxy msg, write :: Proxy msg)`. The reason both rows are supplied is to ensure the message appears regardless of which stream function one calls (e.g. `read/write`).
However, upon writing this, I realize that `Stream.onClose` doesn't verify whether the rows in the Stream are `read` or `write`. Hmm... :thinking:...

For goal 4, we just newtype a Boolean and use `"ipc"` or `undefined` depending on that boolean. Per my reading of the docs, `fork` requires one to include `"ipc"` somewhere in the stdio array, so that function overrides the defaults with `true` and forces the user to return `true` for this option.

For goal 5, the array's first 4 values will always correspond to `[ stdin, stdout, stderr, ipc ]` and each value is either undefined or one of the options (e.g. `inherit`, `pipe`, etc.) described in Node's docs. Thus, anything one wants to add after that will be at the 5th index (i.e. `stdio[4]`) or higher. If one wants to further track things, they will probably newtype `ChildProces` so they can access the corresponding slot in the `stdio` array with type-safety guarantees.

Reply:
> ignore will provide a row of (read :: Proxy "some msg", write :: Proxy "some msg")....
> ...
> However, upon writing this, I realize that `Stream.onClose` doesn't verify whether the rows in the Stream are read or write. Hmm... :thinking: ....

The issue I was raising above is that my previous approach would allow one to use `ignore` and then call `Stream.close (stdout cp)` even though `stdout` in that usage would be `null`. The latest commit updates `ChildProcess` and `StdIO` to store the full type. So, using `ignore` means calling `stdout cp` now returns `Void`.